### PR TITLE
2-to-3: Start with types

### PR DESCRIPTION
**Pull request by grahamc**
A rethinking of #1227 which starts by adding types where there are problems. Adds types and corrects issues where `mypy` was unhappy, after  asking it to check untyped defs.

(Don't merge.)

### CODE REVIEW

File: nixops/deployment.py [hunk 1]
```diff
@@ -137,11 +158,39 @@ def get_machine(self, name):
...
+            raise Exception("Bug: Deployment.definitions['{}'] is None.".format(name))
+        return defn
+
+    def _machine_definition_for_required(
+        self, name: str
+    ) -> nixops.backends.MachineDefinition:
+        defn = self._definition_for_required(name)
+        if not isinstance(defn, nixops.backends.MachineDefinition):
+            raise Exception("Definition named '{}' is not a machine.".format(name))
+        return defn
```
Comment (grahamc) [inline nixops/deployment.py]: Under the old behavior:

1. self.definitions was None, causing `self.definititions[name]` to be an Exception -- no change
2. self.definitions was a Dict, causing `self.definititions[name]` to be possibly None -- generally not causing an Exception, but instead propogating a None around
3. self.definitions was a Dict, `self.definititions[name]` was something, but not a Machine -- instead a different kind of resource. I believe this would cause an exception.

In other words, the only case I'm somewhat unsure about is the None case.

**Comment by grahamc** [comment]
Also: there are commits from the other PR that I'm going to bring over here.

**Comment by worldofpeace** [comment]
Fab :sparkles:

**Comment by grahamc** [comment]
At this point, mypy says this:

```
Mypy Type Check Coverage Summary
================================

Script: index

+--------------------------------+-------------------+----------+
| Module                         | Imprecision       | Lines    |
+--------------------------------+-------------------+----------+
| nixops                         |   0.00% imprecise |    0 LOC |
| nixops.ansi                    |   0.00% imprecise |   18 LOC |
| nixops.backends                |  53.29% imprecise |  516 LOC |
| nixops.backends.none           |  59.83% imprecise |  117 LOC |
| nixops.deployment              |  49.07% imprecise | 1718 LOC |
| nixops.diff                    |  20.00% imprecise |  215 LOC |
| nixops.known_hosts             |  51.39% imprecise |   72 LOC |
| nixops.logger                  |  60.90% imprecise |  156 LOC |
| nixops.nix_expr                |  57.88% imprecise |  349 LOC |
| nixops.parallel                |  28.45% imprecise |  116 LOC |
| nixops.plugins                 |  58.33% imprecise |   12 LOC |
| nixops.plugins.hookspecs       |  32.00% imprecise |   25 LOC |
| nixops.resources               |  49.15% imprecise |  293 LOC |
| nixops.resources.commandOutput |  19.47% imprecise |  113 LOC |
| nixops.resources.ssh_keypair   |  48.33% imprecise |   60 LOC |
| nixops.script_defs             |  62.32% imprecise | 1128 LOC |
| nixops.ssh_util                |  46.33% imprecise |  341 LOC |
| nixops.state                   |  60.00% imprecise |   75 LOC |
| nixops.statefile               |  47.35% imprecise |  264 LOC |
| nixops.util                    |  23.87% imprecise |  507 LOC |
+--------------------------------+-------------------+----------+
| Total                          |  48.56% imprecise | 6095 LOC |
+--------------------------------+-------------------+----------+
```

**Comment by dhess** [comment]
I've done a few deployments and reboots with this PR, and combined them with `include` and `exclude`. It worked for me!

**Comment by grahamc** [comment]
Now improved to:

```
+--------------------------------+-------------------+----------+
| Module                         | Imprecision       | Lines    |
+--------------------------------+-------------------+----------+
| nixops                         |   0.00% imprecise |    0 LOC |
| nixops.ansi                    |   0.00% imprecise |   18 LOC |
| nixops.backends                |  55.25% imprecise |  514 LOC |
| nixops.backends.none           |  59.83% imprecise |  117 LOC |
| nixops.deployment              |  49.53% imprecise | 1712 LOC |
| nixops.diff                    |  18.27% imprecise |  208 LOC |
| nixops.known_hosts             |  11.43% imprecise |   70 LOC |
| nixops.logger                  |  60.90% imprecise |  156 LOC |
| nixops.nix_expr                |  57.88% imprecise |  349 LOC |
| nixops.parallel                |  28.45% imprecise |  116 LOC |
| nixops.plugins                 |  58.33% imprecise |   12 LOC |
| nixops.plugins.hookspecs       |  32.00% imprecise |   25 LOC |
| nixops.resources               |  49.15% imprecise |  293 LOC |
| nixops.resources.commandOutput |  19.47% imprecise |  113 LOC |
| nixops.resources.ssh_keypair   |  48.33% imprecise |   60 LOC |
| nixops.script_defs             |  62.32% imprecise | 1128 LOC |
| nixops.ssh_util                |  46.33% imprecise |  341 LOC |
| nixops.state                   |  22.37% imprecise |   76 LOC |
| nixops.statefile               |  47.35% imprecise |  264 LOC |
| nixops.util                    |  23.87% imprecise |  507 LOC |
+--------------------------------+-------------------+----------+
| Total                          |  47.87% imprecise | 6083 LOC |
+--------------------------------+-------------------+----------+
```

In particular, known_hosts 51.39% -> 11.43%, state 60% -> 22%!

**Comment by aanderse** [comment]
I found no issues with my usual workflow on this branch, using the `none` backend :+1:

**Comment by talyz** [comment]
I didn't run into any issues when running without plugins, trying out the following commands: `create`, `deploy`, `info`, `check`, `ssh`, `list`, `delete`, `export`, `modify` and `list-plugins`.

**Comment by worldofpeace** [comment]
:two_hearts: :two_hearts: :rainbow:
